### PR TITLE
XNIO-195 OSGi support

### PR DESCRIPTION
**Pull request by hwellmann**
Added manifest headers.

Using Declarative Services to inject XnioProvider (non-intrusive
in non-OSGi environment).

### CODE REVIEW

**Comment by ecki** [comment]
What is the FileChannel remove?

**Comment by hwellmann** [comment]
It was duplicate in src/main/java and src/test/java, causing a compiler error.

**Comment by stuartwdouglas** [comment]
This needs to be merged before I can merge the similar OSGI PR for Undertow

**Comment by marcusaram** [comment]
What is the status of this?

File: api/src/main/java/org/xnio/XnioProviderHolder.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * <p>
+ * Intended to be used with OSGi Declarative Services. No side
+ * effects when not running under OSGi.
+ * 
+ * @author Harald Wellmann
+ *
+ */
+public class XnioProviderHolder {
+
+    private static XnioProvider provider;
```
Comment (marschall) [inline api/src/main/java/org/xnio/XnioProviderHolder.java:33]: Should that be volatile?

Comment (hwellmann) [inline api/src/main/java/org/xnio/XnioProviderHolder.java:33]: What problem would `volatile` solve?

**Comment by hwellmann** [comment]
@marcuskara: OSGified versions of XNIO and Undertow exist in the Pax Tipi project - browse group `org.ops4j.pax.tipi` in Maven Central. Of course, I would still prefer the official artifacts to be OSGI-ready.

File: api/src/main/java/org/xnio/XnioProviderHolder.java (hunk 1 above)
Comment (marschall) [inline api/src/main/java/org/xnio/XnioProviderHolder.java:33]: unsafe publication

**Comment by marcusaram** [comment]
@hwellmann Thank you, I'll take a look.

**Comment by arashi01** [comment]
Any update on this?

**Comment by benjumanji** [comment]
Is all that is blocking this in principle the lack of safe publication? I'd be happy to rebase and tidy this up if @hwellmann doesn't object and it were likely to be accepted.

File: api/src/main/java/org/xnio/Xnio.java [hunk 2]
```diff
@@ -235,6 +235,9 @@ private static Xnio doGetInstance(final String provider, final ServiceLoader<Xni
                 msg.debugf(t, "Skipping non-loadable provider");
             }
         }
+        if (XnioProviderHolder.getProvider() != null) {
```
Comment (ecki) [inline api/src/main/java/org/xnio/Xnio.java:238]: can we have that `provider = XnioProiderHolder.getProvider(); if (provider != null return provider.getInstance();`?

File: api/src/main/java/org/xnio/XnioProviderHolder.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * Holds an XnioProvider injected by a provider implementation.
+ * <p>
+ * Intended to be used with OSGi Declarative Services. No side
+ * effects when not running under OSGi.
+ * 
+ * @author Harald Wellmann
+ *
+ */
+public class XnioProviderHolder {
```
Comment (ecki) [inline api/src/main/java/org/xnio/XnioProviderHolder.java:31]: I dont like the fact that this is not implementing the Holder pattern, so is somewhat confusing...

**Comment by benjumanji** [comment]
The other alternative is to add a getInstance method that calls `ServiceLoader.load(XnioProvider.class)`. This combined with some manifest headers would allow OSGi loading with no problems at all (using OSGi extender). Downstream libs would need a toggle and some manifest headers, but  that would be it. No extra static members required.

**Comment by oscerd** [comment]
Any news on this?

**Comment by benjumanji** [comment]
I started looking at how to make this work with spi-fly (which I feel is a better approach than what's in this pull request), but got pulled onto other work, but I plan to come back to it in the next month or so.

**Comment by oscerd** [comment]
Thanks :-)

**Comment by lostiniceland** [comment]
@benjumanji Thanks for picking this up.

Out of curiosity: what would spi-fly solve what this PR (using Declarative Services) would not?

**Comment by gnodet** [comment]
Superseced by https://github.com/xnio/xnio/pull/111

**Comment by gnodet** [comment]
This can be closed now.

**Pull request closed without merging**